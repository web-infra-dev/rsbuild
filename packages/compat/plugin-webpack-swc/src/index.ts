export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
