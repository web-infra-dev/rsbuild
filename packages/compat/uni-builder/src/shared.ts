import { deepmerge } from '@rsbuild/shared/deepmerge';
import type {
  RsbuildPlugin,
  RsbuildConfig as RsbuildRspackConfig,
} from '@rsbuild/core';
import type { RsbuildConfig as RsbuildWebpackConfig } from '@rsbuild/webpack';
import type { UniBuilderRspackConfig, UniBuilderWebpackConfig } from './types';

export function parseCommonConfig<B = 'rspack' | 'webpack'>(
  uniBuilderConfig: B extends 'rspack'
    ? UniBuilderRspackConfig
    : UniBuilderWebpackConfig,
): {
  rsbuildConfig: B extends 'rspack'
    ? RsbuildRspackConfig
    : RsbuildWebpackConfig;
  rsbuildPlugins: RsbuildPlugin[];
} {
  const rsbuildConfig = deepmerge({}, uniBuilderConfig);
  const { output } = rsbuildConfig;

  if (output.cssModuleLocalIdentName) {
    output.cssModules ||= {};
    output.cssModules.localIdentName = output.cssModuleLocalIdentName;
    delete output.cssModuleLocalIdentName;
  }

  return {
    rsbuildConfig: rsbuildConfig,
    rsbuildPlugins: [],
  };
}
