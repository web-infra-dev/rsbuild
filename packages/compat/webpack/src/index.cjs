/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
