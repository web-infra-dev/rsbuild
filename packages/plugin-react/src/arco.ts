import {
  useSSR,
  isPackageInstalled,
  type SharedRsbuildPluginAPI,
} from '@rsbuild/shared';

export const applyArcoSupport = (api: SharedRsbuildPluginAPI) => {
  const ARCO_NAME = '@arco-design/web-react';
  const ARCO_ICON = `${ARCO_NAME}/icon`;

  api.modifyRsbuildConfig((rsbuildConfig) => {
    const { transformImport = [] } = rsbuildConfig.source || {};

    if (
      transformImport === false ||
      !isPackageInstalled(ARCO_NAME, api.context.rootPath)
    ) {
      return;
    }

    const isUseSSR = useSSR(api.context.target);

    if (!transformImport?.some((item) => item.libraryName === ARCO_NAME)) {
      transformImport.push({
        libraryName: ARCO_NAME,
        libraryDirectory: isUseSSR ? 'lib' : 'es',
        camelToDashComponentName: false,
        style: true,
      });
    }

    if (!transformImport?.some((item) => item.libraryName === ARCO_ICON)) {
      transformImport.push({
        libraryName: ARCO_ICON,
        libraryDirectory: isUseSSR ? 'react-icon-cjs' : 'react-icon',
        camelToDashComponentName: false,
      });
    }

    rsbuildConfig.source ||= {};
    rsbuildConfig.source.transformImport = transformImport;
  });
};
