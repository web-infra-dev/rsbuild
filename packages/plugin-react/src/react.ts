import path from 'node:path';
import type { RsbuildConfig, RsbuildPluginAPI, Rspack } from '@rsbuild/core';
import {
  type BundlerChain,
  type ChainIdentifier,
  SCRIPT_REGEX,
  deepmerge,
  isUsingHMR,
} from '@rsbuild/shared';
import type { PluginReactOptions } from '.';

const modifySwcLoaderOptions = ({
  chain,
  CHAIN_ID,
  modifier,
}: {
  chain: BundlerChain;
  CHAIN_ID: ChainIdentifier;
  modifier: (config: Rspack.SwcLoaderOptions) => Rspack.SwcLoaderOptions;
}) => {
  const ruleIds = [CHAIN_ID.RULE.JS, CHAIN_ID.RULE.JS_DATA_URI];

  for (const ruleId of ruleIds) {
    if (chain.module.rules.has(ruleId)) {
      const rule = chain.module.rule(ruleId);
      if (rule.uses.has(CHAIN_ID.USE.SWC)) {
        rule.use(CHAIN_ID.USE.SWC).tap(modifier);
      }
    }
  }
};

export const applyBasicReactSupport = (
  api: RsbuildPluginAPI,
  options: PluginReactOptions,
) => {
  const REACT_REFRESH_PATH = require.resolve('react-refresh');

  api.modifyBundlerChain(async (chain, { CHAIN_ID, isDev, isProd, target }) => {
    const config = api.getNormalizedConfig();
    const usingHMR = isUsingHMR(config, { isProd, target });
    const reactOptions: Rspack.SwcLoaderTransformConfig['react'] = {
      development: isDev,
      refresh: usingHMR,
      runtime: 'automatic',
      ...options.swcReactOptions,
    };

    modifySwcLoaderOptions({
      chain,
      CHAIN_ID,
      modifier: (opts) => {
        const extraOptions: Rspack.SwcLoaderOptions = {
          jsc: {
            parser: {
              syntax: 'typescript',
              // enable supports for React JSX/TSX compilation
              tsx: true,
            },
            transform: {
              react: reactOptions,
            },
          },
        };

        return deepmerge(opts, extraOptions);
      },
    });

    if (!usingHMR) {
      return;
    }

    chain.resolve.alias.set('react-refresh', path.dirname(REACT_REFRESH_PATH));

    const { default: ReactRefreshRspackPlugin } = await import(
      '@rspack/plugin-react-refresh'
    );

    chain
      .plugin(CHAIN_ID.PLUGIN.REACT_FAST_REFRESH)
      .use(ReactRefreshRspackPlugin, [
        {
          include: [SCRIPT_REGEX],
          ...options.reactRefreshOptions,
        },
      ]);
  });
};

export const applyReactProfiler = (api: RsbuildPluginAPI) => {
  api.modifyRsbuildConfig((config, { mergeRsbuildConfig }) => {
    const enableProfilerConfig: RsbuildConfig = {
      output: {
        minify: {
          jsOptions: {
            // Need to keep classnames and function names like Components for debugging purposes.
            mangle: {
              keep_classnames: true,
              keep_fnames: true,
            },
          },
        },
      },
    };
    return mergeRsbuildConfig(config, enableProfilerConfig);
  });

  api.modifyBundlerChain((chain) => {
    // Replace react-dom with the profiling version.
    // Reference: https://gist.github.com/bvaughn/25e6233aeb1b4f0cdb8d8366e54a3977
    chain.resolve.alias.set('react-dom$', 'react-dom/profiling');
    chain.resolve.alias.set('scheduler/tracing', 'scheduler/tracing-profiling');
  });
};
