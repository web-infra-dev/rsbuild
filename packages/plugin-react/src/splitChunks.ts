import type { RsbuildPluginAPI, SplitChunks } from '@rsbuild/core';
import { createCacheGroups } from '@rsbuild/shared';
import type { SplitReactChunkOptions } from '.';

const isPlainObject = (obj: unknown): obj is Record<string, any> =>
  obj !== null &&
  typeof obj === 'object' &&
  Object.prototype.toString.call(obj) === '[object Object]';

export const applySplitChunksRule = (
  api: RsbuildPluginAPI,
  options: SplitReactChunkOptions = {
    react: true,
    router: true,
  },
): void => {
  api.modifyBundlerChain((chain, { environment }) => {
    const config = api.getNormalizedConfig({ environment });
    if (config.performance.chunkSplit.strategy !== 'split-by-experience') {
      return;
    }

    const currentConfig = chain.optimization.splitChunks.values();
    if (!isPlainObject(currentConfig)) {
      return;
    }

    const extraGroups: Record<string, (string | RegExp)[]> = {};

    if (options.react) {
      extraGroups.react = [
        'react',
        'react-dom',
        'scheduler',
        ...(process.env.NODE_ENV === 'production'
          ? []
          : ['react-refresh', /@rspack[\\/]plugin-react-refresh/]),
      ];
    }

    if (options.router) {
      extraGroups.router = [
        'react-router',
        'react-router-dom',
        'history',
        /@remix-run[\\/]router/,
      ];
    }

    if (!Object.keys(extraGroups).length) {
      return;
    }

    chain.optimization.splitChunks({
      ...currentConfig,
      cacheGroups: {
        ...(currentConfig as Exclude<SplitChunks, false>).cacheGroups,
        ...createCacheGroups(extraGroups),
      },
    });
  });
};
