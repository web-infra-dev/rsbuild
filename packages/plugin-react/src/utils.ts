import fs from 'fs';
import semver from '@rsbuild/shared/semver';
import { findUp } from '@rsbuild/shared';

export const isBeyondReact17 = async (cwd: string) => {
  const pkgPath = await findUp({ cwd, filename: 'package.json' });

  if (!pkgPath) {
    return false;
  }

  const pkgInfo = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
  const deps = {
    ...pkgInfo.devDependencies,
    ...pkgInfo.dependencies,
  };

  if (typeof deps.react !== 'string') {
    return false;
  }

  return semver.satisfies(semver.minVersion(deps.react)!, '>=17.0.0');
};
