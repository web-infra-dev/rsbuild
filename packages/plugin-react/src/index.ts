import type { RsbuildPlugin } from '@rsbuild/core';
import { applyAntdSupport } from './antd';
import { applyArcoSupport } from './arco';
import { applySplitChunksRule } from './splitChunks';
import { applyBasicReactSupport } from './react';

export { isBeyondReact17 } from './utils';

export const pluginReact = (): RsbuildPlugin => ({
  name: 'rsbuild:react',

  pre: ['rsbuild:swc'],

  setup(api) {
    if (api.context.bundlerType === 'rspack') {
      applyBasicReactSupport(api);
    }
    applyAntdSupport(api);
    applyArcoSupport(api);
    applySplitChunksRule(api);
  },
});
