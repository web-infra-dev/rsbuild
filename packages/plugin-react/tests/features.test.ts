import { expect, describe, it, vi } from 'vitest';
import { pluginReact } from '../src';
import { createStubRsbuild } from '@rsbuild/test-helper';
import { SCRIPT_REGEX } from '@rsbuild/shared';

vi.mock('@rsbuild/shared', async (importOriginal) => {
  const mod = await importOriginal<any>();
  return {
    ...mod,
    isPackageInstalled: () => true,
  };
});

describe('splitChunks', () => {
  it('should apply antd/semi/... splitChunks rule when pkg is installed', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {
        performance: {},
      },
    });

    rsbuild.addPlugins([pluginReact()]);

    const config = await rsbuild.unwrapConfig();

    expect(config.optimization.splitChunks).toMatchSnapshot();
  });

  it('should not apply splitChunks rule when strategy is not split-by-experience', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {
        performance: {
          chunkSplit: {
            strategy: 'single-vendor',
          },
        },
      },
    });

    rsbuild.addPlugins([pluginReact()]);

    const config = await rsbuild.unwrapConfig();

    expect(config.optimization.splitChunks).toMatchSnapshot();
  });
});

describe('transformImport', () => {
  it('should apply antd & arco transformImport', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {},
    });

    rsbuild.addPlugins([pluginReact()]);

    const config = await rsbuild.unwrapConfig();

    expect(
      config.module.rules.find(
        (r) => r.test.toString() === SCRIPT_REGEX.toString(),
      ),
    ).toMatchSnapshot();
  });

  it('should not apply antd & arco when transformImport is false', async () => {
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {
        source: {
          transformImport: false,
        },
      },
    });

    rsbuild.addPlugins([pluginReact()]);

    const config = await rsbuild.unwrapConfig();

    expect(
      config.module.rules.find(
        (r) => r.test.toString() === SCRIPT_REGEX.toString(),
      ),
    ).toMatchSnapshot();
  });
});
