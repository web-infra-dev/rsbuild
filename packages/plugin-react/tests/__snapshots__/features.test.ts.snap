// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`splitChunks > should apply antd/semi/... splitChunks rule when pkg is installed 1`] = `
{
  "cacheGroups": {
    "lib-axios": {
      "name": "lib-axios",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
    },
    "lib-lodash": {
      "name": "lib-lodash",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
    },
    "lib-polyfill": {
      "name": "lib-polyfill",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(tslib\\|core-js\\|@babel\\\\/runtime\\|@swc\\\\/helpers\\)\\[\\\\\\\\/\\]/,
    },
    "lib-react": {
      "name": "lib-react",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react\\|react-dom\\|scheduler\\|react-refresh\\|@rspack\\[\\\\\\\\/\\]plugin-react-refresh\\)\\[\\\\\\\\/\\]/,
    },
    "lib-router": {
      "name": "lib-router",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(react-router\\|react-router-dom\\|history\\|@remix-run\\[\\\\\\\\/\\]router\\)\\[\\\\\\\\/\\]/,
    },
  },
  "chunks": "all",
  "enforceSizeThreshold": 50000,
}
`;

exports[`splitChunks > should apply splitChunks.react/router plugin option when strategy is split-by-experience 1`] = `
{
  "cacheGroups": {
    "lib-axios": {
      "name": "lib-axios",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(axios\\|axios-\\.\\+\\)\\[\\\\\\\\/\\]/,
    },
    "lib-lodash": {
      "name": "lib-lodash",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(lodash\\|lodash-es\\)\\[\\\\\\\\/\\]/,
    },
    "lib-polyfill": {
      "name": "lib-polyfill",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]\\(tslib\\|core-js\\|@babel\\\\/runtime\\|@swc\\\\/helpers\\)\\[\\\\\\\\/\\]/,
    },
  },
  "chunks": "all",
  "enforceSizeThreshold": 50000,
}
`;

exports[`splitChunks > should not apply splitChunks rule when strategy is not split-by-experience 1`] = `
{
  "cacheGroups": {
    "singleVendor": {
      "chunks": "all",
      "enforce": true,
      "name": "vendor",
      "priority": 0,
      "reuseExistingChunk": true,
      "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
    },
  },
  "chunks": "all",
  "enforceSizeThreshold": 50000,
}
`;

exports[`transformImport > should apply antd & arco transformImport 1`] = `
{
  "include": [
    {
      "and": [
        "<ROOT>",
        {
          "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      ],
    },
    /\\\\\\.\\(ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
  ],
  "test": /\\\\\\.\\(js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
  "type": "javascript/auto",
  "use": [
    {
      "loader": "builtin:swc-loader",
      "options": {
        "env": {
          "coreJs": "3.32",
          "mode": "usage",
          "shippedProposals": true,
          "targets": [
            "chrome >= 87",
            "edge >= 88",
            "firefox >= 78",
            "safari >= 14",
          ],
        },
        "isModule": "unknown",
        "jsc": {
          "externalHelpers": true,
          "parser": {
            "decorators": true,
            "syntax": "typescript",
            "tsx": true,
          },
          "preserveAllComments": true,
          "transform": {
            "decoratorMetadata": true,
            "legacyDecorator": true,
            "react": {
              "development": true,
              "refresh": true,
              "runtime": "automatic",
            },
          },
        },
        "rspackExperiments": {
          "import": [
            {
              "camelToDashComponentName": false,
              "libraryDirectory": "es",
              "libraryName": "@arco-design/web-react",
              "style": true,
            },
            {
              "camelToDashComponentName": false,
              "libraryDirectory": "react-icon",
              "libraryName": "@arco-design/web-react/icon",
            },
          ],
        },
        "sourceMaps": true,
      },
    },
  ],
}
`;

exports[`transformImport > should not apply antd & arco when transformImport is false 1`] = `
{
  "include": [
    {
      "and": [
        "<ROOT>",
        {
          "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
        },
      ],
    },
    /\\\\\\.\\(ts\\|tsx\\|jsx\\|mts\\|cts\\)\\$/,
  ],
  "test": /\\\\\\.\\(js\\|jsx\\|mjs\\|cjs\\|ts\\|tsx\\|mts\\|cts\\)\\$/,
  "type": "javascript/auto",
  "use": [
    {
      "loader": "builtin:swc-loader",
      "options": {
        "env": {
          "coreJs": "3.32",
          "mode": "usage",
          "shippedProposals": true,
          "targets": [
            "chrome >= 87",
            "edge >= 88",
            "firefox >= 78",
            "safari >= 14",
          ],
        },
        "isModule": "unknown",
        "jsc": {
          "externalHelpers": true,
          "parser": {
            "decorators": true,
            "syntax": "typescript",
            "tsx": true,
          },
          "preserveAllComments": true,
          "transform": {
            "decoratorMetadata": true,
            "legacyDecorator": true,
            "react": {
              "development": true,
              "refresh": true,
              "runtime": "automatic",
            },
          },
        },
        "sourceMaps": true,
      },
    },
  ],
}
`;
